File: benchmarks/chaostools/maps.jl
using DynamicalSystemsBase, ChaosTools, StaticArrays

include(joinpath(@__DIR__, "_common.jl"))

# Tier 0.2/0.3: same params as benchmarks/lyapax/maps.py.

logistic_rule(u, p, n) = SVector(4.0 * u[1] * (1.0 - u[1]))
u0_logistic = [0.4]
ds = DeterministicIteratedMap(logistic_rule, u0_logistic, nothing)
run_fn = () -> lyapunovspectrum(ds, 500_000, 1; u0=u0_logistic, Δt=1, Ttr=1000)
first_s, warm_s, exponents = time_and_run(run_fn)
emit("chaostools", "logistic_map_tier0.2", exponents, first_s, warm_s)

tent_rule(u, p, n) = SVector(u[1] < 0.5 ? 2.0 * u[1] : 2.0 * (1.0 - u[1]))
u0_tent = [0.4]
ds = DeterministicIteratedMap(tent_rule, u0_tent, nothing)
run_fn = () -> lyapunovspectrum(ds, 500_000, 1; u0=u0_tent, Δt=1, Ttr=1000)
first_s, warm_s, exponents = time_and_run(run_fn)
emit("chaostools", "tent_map_tier0.2", exponents, first_s, warm_s)

function henon_rule(u, p, n)
    a, b = p
    return SVector(1.0 - a * u[1]^2 + u[2], b * u[1])
end
u0_henon = [0.1, 0.1]
p_henon = [1.4, 0.3]
ds = DeterministicIteratedMap(henon_rule, u0_henon, p_henon)
run_fn = () -> lyapunovspectrum(ds, 200_000, 2; u0=u0_henon, Δt=1, Ttr=1000)
first_s, warm_s, exponents = time_and_run(run_fn)
emit("chaostools", "henon_map_tier0.3", exponents, first_s, warm_s)

File: benchmarks/lyapax/maps.py
"""Tier 0.2/0.3: logistic map, tent map, Henon map. Same params as
tests/test_lyapunov_core.py's map tests.
"""
import jax.numpy as jnp

from lyapax.core import ODEProblem, lyapunov_spectrum
from lyapax import systems

from _common import time_and_run, emit


def run_logistic():
    step = systems.logistic_map(r=4.0)
    problem = ODEProblem(step_fn=step, state0=jnp.array([0.4]), dt=1.0)
    return lyapunov_spectrum(problem, n_steps=500_000, renorm_every=1, t_transient=1_000.0)


def run_tent():
    step = systems.tent_map()
    problem = ODEProblem(step_fn=step, state0=jnp.array([0.4]), dt=1.0)
    return lyapunov_spectrum(problem, n_steps=500_000, renorm_every=1, t_transient=1_000.0)


def run_henon():
    step = systems.henon_map(a=1.4, b=0.3)
    problem = ODEProblem(step_fn=step, state0=jnp.array([0.1, 0.1]), dt=1.0)
    return lyapunov_spectrum(problem, n_steps=200_000, renorm_every=1, t_transient=1_000.0)


if __name__ == "__main__":
    for name, fn in [("logistic_map_tier0.2", run_logistic),
                      ("tent_map_tier0.2", run_tent),
                      ("henon_map_tier0.3", run_henon)]:
        first_s, warm_s, result = time_and_run(fn)
        emit("lyapax", name, result.exponents, first_s, warm_s)
